src/Components/Components/src/Rendering/RendererSynchronizationContextDispatcher.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.Rendering
+{
+    internal class RendererSynchronizationContextDispatcher : Dispatcher
+    {
+        private readonly RendererSynchronizationContext _context;
+
+        public RendererSynchronizationContextDispatcher()
+        {
+            _context = new RendererSynchronizationContext();
+        }
+
+        public override event UnhandledExceptionEventHandler UnhandledException
+        {
+            add
+            {
+                _context.UnhandledException += value;
+            }
+            remove
+            {
+                _context.UnhandledException -= value;
+            }
+        }
+
+        public override bool CheckAccess() => SynchronizationContext.Current == _context;
```

Comment:
Separate from this PR - I suspect that this *optimization* is actually incorrect. You might have the sync context set as *your* sync context, but it doesn't prevent work from being scheduled on the sync context concurrent with that you're currently executing.

Reply:
TBH I don't understand your comment. Perhaps the method name `CheckAccess` is too ambiguous really. What exactly is it checking "access" to, and what does "access" imply? The XML docs say:

>  Returns a value that determines whether using the dispatcher to invoke a work item is required from the current context.

Sounds like a more explicit method name would be something like `IsDispatcherInvocationRequired`.

> it doesn't prevent work from being scheduled on the sync context concurrent with that you're currently executing.

This is the bit I'm not getting. Why would it be a problem for some other code to schedule some work on the sync context at the same time? We'll still run them in series, so I don't know why it's a problem.

Reply:
> This is the bit I'm not getting. Why would it be a problem for some other code to schedule some work on the sync context at the same time? We'll still run them in series, so I don't know why it's a problem.

The implication of `CheckAccess()` is that you already have access, so there's no need to *schedule*.

I'm saying that I suspect this isn't implemented correctly - I don't think the test being done here is robust enough.